Revise the explanation of the cartridge header

ihonen opened the pull request:
I clarified, reworded and reformatted a bunch of stuff.

pinobatch commented on src/The_Cartridge_Header.md:
```diff
@@ -80,34 +80,34 @@ Code | Publisher
+`38` | Hudson Soft
 `39` | Banpresto
-`41` | Ubi Soft
+`41` | Ubisoft
```
Point of style: Ubisoft was known as Ubi Soft prior to 2003 according to [its article on Logopedia](https://logos.fandom.com/wiki/Ubisoft). Should we use the modern spelling or the GB era spelling?

ihonen replied:
That's a good question, I was thinking about that as well. I went through every single company in those two lists and Ubisoft is not the only one that is subject to this issue.

I would argue that, since the name is not stored as a verbatim string in the game binary, it makes sense to refer to the entity with its present name, but I can see the other point of view as well. And it does bother me that this solution is not entirely technically correct, which is the best kind of correct. 😛

ISSOtm replied:
Given that the company that was licensed and that published the game was what it was back then, I'll advocate for the old style.

aaaaaa123456789 replied:
"Ubisoft (Ubi Soft until 2003-09-09)"

avivace reviewed (changes requested):
Looks good to me, but I'd avoid changing the publisher names, or at least we should do it (and discuss about it) in a different PR. Can you revert the changes on the publisher names?

ihonen commented:
> Looks good to me, but I'd avoid changing the publisher names, or at least we should do it (and discuss about it) in a different PR. Can you revert the changes on the publisher names?

That's probably a good idea. I think I'll revert the spelling changes and open another PR for that.

ihonen commented:
I made a very slight amendment to the first commit and saved the others for closer inspection afterwards. I'm unfamiliar with the pull request system on GitHub, so I'm sorry if I made a mess of this – I have no idea what I'm doing.

ISSOtm commented:
PRs work by trying to add any changes reachable from the "source" branch (here, `ihonen/master`) but not from the "target" branch (here, `gbdev/master`). Put differently: it'll try to apply the commits from `ihonen/master` that aren't in `gbdev/master`.

So tl;dr, yes, you must create another branch to make a separate PR.

Also, don't worry about potentially muddying the history of the branches you're trying to PR, as maintainers automatically get rights to modify them, and can clean that up right before merging :)

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
 
-An internal information area is located at $0100-014F in each cartridge.
-It contains the following values:
+Each cartridge contains a header, located at the addresses `$0100`-`$014F`.
```
```suggestion
Each cartridge contains a header, located at the address range `$0100`-`$014F`.
```

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
-An internal information area is located at $0100-014F in each cartridge.
-It contains the following values:
+Each cartridge contains a header, located at the addresses `$0100`-`$014F`.
+The cartridge header contains the following information:
```
```suggestion
The cartridge header provides the following information:
```

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
-in the cartridge. Usually this 4 byte area contains a `nop` instruction,
-followed by a `jp $0150` instruction. But not always.
+After displaying the Nintendo logo, the built-in boot procedure jumps to
+this address (`$0100`), which should then jump to the actual main program
```
```suggestion
to address `$0100`, which should then jump to the actual main program
```

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
-incorrect. A CGB verifies only the first half ($18 bytes of) the bitmap, but
-others (for example a Game Boy pocket) verify all $30 bytes, as does the
+
+The Game Boy's boot procedure **first displays** the logo and **then verifies**
```
why the bold usage here?

ihonen replied:
I found the fact that the logo is _first_ displayed and only _then_ verified a bit surprising, so I guess I wanted to emphasize that. But I agree with @ISSOtm's comment - there's no point in going into the details of the power-up sequence in this document, since that is outside its scope. @ISSOtm would you mind making the change?

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
+ROM **locks itself up**.
+
+A CGB verifies only the first half (`$18` bytes) of the bitmap, but other models
+(the Game Boy Pocket, for example) verify all `$30` bytes, as does the
```
```suggestion
(e.g. the Game Boy Pocket) verify all `$30` bytes, as does the
```

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
+
+A CGB verifies only the first half (`$18` bytes) of the bitmap, but other models
+(the Game Boy Pocket, for example) verify all `$30` bytes, as does the
 Game Boy Advance.
```
This part is quite unclear. Other models (like game pocket) verify all bytes, like the GBA? Doesn't sound okay

CasualPokePlayer replied:
The GBA does not check all bytes actually (which can be seen in the disassembly https://github.com/ISSOtm/gb-bootroms/blob/master/src/cgb.asm).

avivace replied:
Thanks for the heads up @CasualPokePlayer. Link is a bit problematic as the source is leaked material. Maybe we can just link to a specific line or give some context and a link instead of copying that information here?

CasualPokePlayer replied:
https://www.youtube.com/watch?v=0dJrYeRbIgc there's also this video showing the GBA only doing the half logo check.

avivace replied:
> https://www.youtube.com/watch?v=0dJrYeRbIgc there's also this video showing the GBA only doing the half logo check.

This may be good enough for now. Maybe we can prepare a test rom?

CasualPokePlayer replied:
Better wording here might just be "CGB and future devices only check the first half, while all pre-CGB devices check the full logo." Although an interesting point in that is on the SGB, while the full logo is checked, it is not be the bootrom. The bootrom simply sends the logo to the SNES software (as can be seen in the diassembly in that repo), which then that SNES software verifies the logo (and refuses to boot if it's wrong).

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
-Nintendo has reduced the length of this area to 15 characters, and some
-months later they had the fantastic idea to reduce it to 11 characters
-only. The new meaning of the ex-title bytes is described below.
+These bytes contain the title of the game in UPPER CASE ASCII.
```
```suggestion
These bytes contain the title of the game in upper case ASCII.
```

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
-## 0143 - CGB Flag
+In older cartridges these bytes were part of the Title (see above).
+In newer cartridges they contain a 4-character manufacturer code
+(uppercase). The exact purpose of the code is unknown.
```
```suggestion
(uppercase). The purpose is unknown.
```

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
+## 0143 – CGB Flag
+
+In older cartridges this byte was part of the Title (see above). In
+CGB cartridges the most significant bit is used to enable CGB functions. This is
```
Why is this "required"? It sounds like what an "enable" does anyway

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
+CGB cartridges the most significant bit is used to enable CGB functions. This is
+required; otherwise the CGB switches itself into Non-CGB-Mode.
+
+Typical values are:
```
```suggestion
Possible values are:
```

ihonen replied:
What @ISSOtm said - there are other possible values.

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
+Values with bit 7 set, and either bit 2 or 3 set, will switch the
 Game Boy into a special non-CGB-mode called "PGB mode".
 
 TODO: research and document PGB modes...
```
This should go into a "Research needed" box

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
-games are using the header entry at `$014B` instead.
+These bytes specify a two-character ASCII licensee code indicating the
+developer/publisher of the game. The New Licensee Code is present in newer
+cartridges only (those released after the SGB was invented).
```
```suggestion
cartridges only (those released after the SGB was released).
```

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
+## 0146 – SGB Flag
 
-Specifies whether the game supports SGB functions, common values are:
+This byte specifies whether the game supports SGB functions.
```
Why repeat "this byte"? we're explaining it anyway..

ISSOtm replied:
I agree the repetition isn't all that great, but I think it's acceptable to be more precise, and can be less monotonous by alternative with "this area" and/or "this field".

ihonen replied:
I think in a technical text precision is preferable to elegance, and I simply thought that "this byte" is a consistent way to format things.

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
+## 0149 – RAM Size
 
-Specifies the size of the external RAM in the cartridge (if any).
+This byte specifies the size of the RAM on the cartridge (if any).
```
Again, there's no need of repeating "this byte" on every pararaph..

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
-Contains a 16 bit checksum (upper byte first) across the whole cartridge
-ROM. Produced by adding all bytes of the cartridge (except for the two
-checksum bytes). The Game Boy doesn't verify this checksum.
+The boot ROM does not verify this checksum.
```
Why the change from "Game Boy" to boot rom?

ISSOtm replied:
Because it's more precise, honestly. Though you are right that for a general negation, the suggested passive voice is better.

ihonen replied:
Again what @ISSOtm said - just stating that the checksum is verified by the Game Boy is pretty ambiguous. While the exact details of the boot procedure are outside the scope of this document, I don't think it's bad to simply mention that it is the boot ROM that does the checksum checking.

avivace commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
-Contains a 16 bit checksum (upper byte first) across the whole cartridge
-ROM. Produced by adding all bytes of the cartridge (except for the two
-checksum bytes). The Game Boy doesn't verify this checksum.
+The boot ROM does not verify this checksum.
```
```suggestion
This checksum is not verified.
```

ihonen replied:
I think referring to the boot ROM in this paragraph as well reduces the likelihood of any kind of confusion, so I would prefer to keep the former wording.

CasualPokePlayer replied:
If it's worth a mention, the Game Boy Tower actually verifies this checksum.

ISSOtm replied:
```suggestion
This checksum is not verified, except by the GB Tower's boot ROM.
```

Do you have a source we can quote?

CasualPokePlayer replied:
The GB Tower's boot ROM does not verify it (in fact it is a literal copy of the MGB bootrom, or otherwise its peculiar CGB bootrom in Stadium 2 for the GBC games). It's when it's loading in the ROM file (presumably error handling in case transfer pak here failed). It's not particularly hard to check (given an emulator able to handle it) or even just real hardware (which is what I checked).

ihonen replied:
@ISSOtm If you could resolve this final issue then I think this PR will be ready for merge. :)

avivace reviewed (changes requested):
Thanks! Overall, looks like a nice improvements, but there are some things that still need a bit of discussion (I noted both the ones changed/introduced by you and unchanged ones).

ihonen commented:
> PRs work by trying to add any changes reachable from the "source" branch (here, `ihonen/master`) but not from the "target" branch (here, `gbdev/master`). Put differently: it'll try to apply the commits from `ihonen/master` that aren't in `gbdev/master`.
> 
> So tl;dr, yes, you must create another branch to make a separate PR.
> 
> Also, don't worry about potentially muddying the history of the branches you're trying to PR, as maintainers automatically get rights to modify them, and can clean that up right before merging :)

Cheers, that was helpful. I'm familiar with Git but pretty much exclusively on the command line, so this fancy pull request system is all new and confusing to me. :)

ihonen commented:
Any update in regards to this?

ISSOtm commented:
What do you think of the review comments above?

ISSOtm commented:
(...not sure why Ctrl+Enter defaulted to Comment+Close instead of Comment, but okay.)

ihonen commented:
> What do you think of the review comments above?

I already addressed most of them and implemented the ones I agreed with, but I guess it wasn't that obvious to everyone else since I amended the original commit instead of making new ones.

avivace commented:
> > What do you think of the review comments above?
> 
> I already addressed most of them and implemented the ones I agreed with, but I guess it wasn't that obvious to everyone else since I amended the original commit instead of making new ones.

@ihonen , if you take a look at the top of this page, there is a series of threads started from single lines. Can you jump there and reply to the single points? E.g.:
![image](https://user-images.githubusercontent.com/14352721/176530458-25907b89-00e2-49ea-b043-1fb714582fae.png)

ISSOtm commented:
Note that you may want to use the `Add suggestion to batch` button to apply several of them at once.

ihonen commented:
> Note that you may want to use the Add suggestion to batch button to apply several of them at once.

Thanks, I didn't notice that option.

ihonen commented:
I kind of messed this whole thing up by amending my original commit (and this was two weeks ago). I don't think the pull request UI can really account for that (I find everything above confusing after the amendment because I don't remember which suggestions exactly I amended the original commit with), so I think the easiest way to resolve this is to remove the earlier discussion and re-review the changes. I'll know better next time.

avivace commented:
> I kind of messed this whole thing up by amending my original commit (and this was two weeks ago). I'm not sure the UI can really account for that (I find everything above confusing after the amendment because I don't remember which suggestions exactly I amended the original commit with), so I think the easiest way to resolve this is to remove the earlier discussion and re-review the changes. Next time I'll know better.
> 
> I mean, just to get this thing merged, it'd be easiest and least confusing to start fresh.

@ihonen Again, please, do take a look at the number of threads that were open. You can see the full list just by scrolling up. No matter what you did, github is generally able to understand which of the "commented" lines were changed, and in case, it marks them as "outdated" so we can consider those as "resolved". Yet, there are some of them that weren't changed, and even if github UI messes it up somewhat, you have every tool to go through them "manually" and checking what you did for the single lines in your commit, so why not do that?

E.g.:
![image](https://user-images.githubusercontent.com/14352721/176999713-e4a346f6-041a-4455-81b0-ebdb0accc319.png)
It's enough to CTRL+F the part in red to see that this is still in your commit, so this still needs to be handled. Please, request another review after you took a look at the points we already raised.

ihonen commented:
@avivace I have responded to multiple comments disagreeing with the comment but haven't got any further response, so I don't know what you want me to do. In any case, I have now accepted all changes I agree with and I suggest we get done with this pull request because it's getting ever messier. We can perfectly well make adjustments to the text later, it's not like the text is getting worse with this merge and it's also not like any text is ever finished per se.

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,75 @@
-publisher of the game. These two bytes are used in newer games only
-(games that have been released after the SGB has been invented). Older
-games are using the header entry at `$014B` instead.
+These bytes specify a two-character ASCII licensee code indicating the
+developer/publisher of the game. The New Licensee Code is present in newer
+cartridges only (those released after the SGB was released).
+Older cartridges use the Old Licensee Code at `$014B` instead.
```
```suggestion
This area contains a two-character ASCII "licensee code" indicating the game's publisher.
It is only meaningful if the [Old Licensee Code](<#014B – Old Licensee Code>) is exactly $33 (which is essentially all games made after the SGB was released); otherwise, the old code must be considered.
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
 
-- `$00`: No SGB functions (Normal Game Boy or CGB only game)
-- `$03`: Game supports SGB functions
+Common values are:
+
+Code  | SGB support
+------|-----------------------------------------------------
+`$00` | No SGB functions (regular Game Boy or CGB-only game)
+`$03` | The game supports SGB functions
 
 The SGB disables its SGB functions if this byte is set to a value other than `$03`.
```
```suggestion
The SGB will ignore any [packets](<#Command Packet Transfers>) if this byte is set to a value other than $03 (which is typically $00).
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
 
-Specifies whether the game supports SGB functions, common values are:
+This byte specifies whether the game supports SGB functions.
 
```
```suggestion
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
- $FF | HuC1+RAM+BATTERY
+## 0147 – Cartridge Type
+
+This byte indicates what kind of Memory Bank Controller (and other hardware)
+the cartridge contains.
```
```suggestion
This byte indicates what kind of hardware is present on the cartridge, notably its [mapper](#MBCs).
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
- $52 | 1.1 MByte | 72 [^weird_rom_sizes]
- $53 | 1.2 MByte | 80 [^weird_rom_sizes]
- $54 | 1.5 MByte | 96 [^weird_rom_sizes]
+MBC3 with a 64 KiB RAM refers to MBC30, used only in
+_Pocket Monsters: Crystal Version_ (the Japanese version of
+_Pokémon Crystal Version_).
```
```suggestion
MBC3 with 64 KiB of SRAM refers to MBC30, used only in _Pocket Monsters: Crystal Version_ (the Japanese version of _Pokémon Crystal Version_).
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
 
-Specifies the size of the external RAM in the cartridge (if any).
+This byte specifies the size of the RAM on the cartridge (if any).
 
```
```suggestion

If the [cartridge type](<#0147 – Cartridge Type>) does not include "RAM" in its name, this should be set to 0.
This includes MBC2, since its 512 × 4 bits of memory are built directly into the mapper.

```
...and then the footnote below would be removed.
Then, we'd also delete the MBC2 footnote.

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
-in the filename) are known to use the $01 RAM Size tag, but this is believed
+Various "PD" ROMs ("Public Domain" homebrew ROMs, generally tagged with `(PD)`
+in the filename) are known to use the `$01` RAM Size tag, but this is believed
 to have been a mistake with early homebrew tools and the PD ROMs often don't use
```
```suggestion
to have been a mistake with early homebrew tools, and the PD ROMs often don't use
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
-- $01: Non-Japanese
+Code  | Destination
+------|------------
+`$00` | Japan
+`$01` | Other
```
```suggestion
`$00` | Japan (and possibly overseas)
`$01` | Overseas only
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
-$33 signals that the New Licensee Code (in header bytes \$0144-0145) is
-used instead. (Super Game Boy functions won't work if \<\> \$33.) A list
-of licensee codes can be found
+This byte is used in older (pre-SGB) cartridges to specify the developer/publisher
+of the game. The value `$33` indicates that the New Licensee Code (in header bytes
+`$0144`-`$0145`) is used instead. (Super Game Boy functions will only work if the
+value is `$33`.)
```
```suggestion
This byte is used in older (pre-SGB) cartridges to specify the game's publisher.
However, the value $33 indicates that the [New Licensee Code](<#0144-0145 – New Licensee Code>) must be considered instead.
(The Super Game Boy will ignore packets unless this value is $33.)
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
+`$0144`-`$0145`) is used instead. (Super Game Boy functions will only work if the
+value is `$33`.)
+
+A list of Old Licensee Codes can be found
 [here](https://raw.githubusercontent.com/gb-archive/salvage/master/txt-files/gbrom.txt).
```
```suggestion
Here is [a list of Old Licensee Codes](https://raw.githubusercontent.com/gb-archive/salvage/master/txt-files/gbrom.txt).
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +137,160 @@ Code | Publisher
+## 014C – Mask ROM Version number
 
-Specifies the version number of the game. That is usually $00.
+This byte specifies the version number of the game. The value is usually `$00`.
```
```suggestion
This byte specifies the version number of the game.
It is usually $00.
```

ISSOtm commented on src/The_Cartridge_Header.md:
````diff
@@ -127,145 +137,160 @@ Code | Publisher
+`$0134`-`$014C`. The boot ROM computes the checksum as follows:
 
+```c
+checksum := 0
+address  := 0x0134
+while address <= 0x014C:
+	checksum := checksum - [address] - 1
+	address  := address + 1
````
```suggestion
uint8_t checksum = 0;
for (uint16_t address = 0x0134; address < 0x014C; address++) {
    checksum = checksum - rom[address] - 1;
}
```
If we want it highlight it as C, let's make it proper C! :stuck_out_tongue:

ISSOtm reviewed (changes requested):
Thanks a lot for the improvements! I only have a few minor changes in mind.

ISSOtm commented:
> I have responded to multiple comments (including the one you refer to above) disagreeing with the comment but haven't got any further response, so I don't know what you want me to do.

It's weird, I don't see any replies from you. Given that it's the default behavior, my guess is that you clicked on the green "Add review comment" button, which actually batches all of the comments in a "review" like I just published, but they aren't published until you click the green "review" button at the top-right of the Files view.

To avoid this, you can either reply to discussions in the Conversation tab of the PR, or click the "Add single comment" when making replies from the Files tab. I agree it's not the most intuitive behavior... from our point of view, you haven't replied to anything—hence avivace thought that this PR had gone inactive. Sorry about that!

ihonen commented:
@ISSOtm Uhhh... And I already found the UI appalling. Thanks for the explanation, I guess that's why it says "Pending" in yellow next to all of my replies. As far as I can tell, I have written all the replies while in the Conversation tab, though. Sorry for making such a mess in any case, I just find the UI _extremely_ confusing and not at all intuitive to use. I'll have to figure it out.

I'll try to get back and clean up this mess to whatever extent I can in the next couple of days when I find the time.

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,71 @@
-incorrect. A CGB verifies only the first half ($18 bytes of) the bitmap, but
-others (for example a Game Boy pocket) verify all $30 bytes, as does the
-Game Boy Advance.
 
```
Adding an explanation of how the thing is encoded might be a good idea?

```suggestion

The way pixels are encoded is as follows: ([more visual aid](https://github.com/ISSOtm/gb-bootroms/blob/2dce25910043ce2ad1d1d3691436f2c7aabbda00/src/dmg.asm#L259-L269))

- $0104–011B encode the top half of the logo, $011C–0133 encode the bottom half of it.
- For each half, each nibble encodes 4 pixels (the MSB corresponds to the leftmost pixel, the LSB to the rightmost); a pixel is lit if its bit is set.
- The 4-pixel "groups" are laid out top to bottom, left to right.
- And finally, the monochrome models upscale the entire thing by a factor of 2 (leading to somewhat chunky pixels).

```

ihonen replied:
Is the logo encoded in a different manner than regular tiles/sprites? I don't remember off the top of my head, but basically what I am asking is that could the explanation be replaced with a link to the pertinent section of the PPU documentation?

ISSOtm replied:
Yes, it is encoded differently. https://github.com/ISSOtm/gb-bootroms/blob/2dce25910043ce2ad1d1d3691436f2c7aabbda00/src/dmg.asm#L221

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,71 @@
-Game Boy Advance.
 
-## 0134-0143 - Title
+The Game Boy's boot procedure **first displays** the logo and **then verifies**
+its contents. If the contents do not match the values above, the boot
+ROM **locks itself up**.
```
```suggestion
The Game Boy's boot procedure [first displays the logo and then checks](<#bypass>) that it matches the dump above.
If it doesn't, the boot ROM **locks itself up**.
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,71 @@
+`$80` | The game supports CGB enhancements, but is backwards compatible with monochrome Game Boys
+`$C0` | The game works on CGB only (the hardware ignores bit 6, so this really functions the same as `$80`)
+
+Values with bit 7 set, and either bit 2 or 3 set, will switch the
 Game Boy into a special non-CGB-mode called "PGB mode".
```
```suggestion
Values with bit 7 set, and either bit 2 or 3 set, will switch the Game Boy into a special non-CGB-mode called "PGB mode".
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
+
+## 0148 – ROM Size
+
+This byte specifies the size of the ROM on the cartridge. In most cases the ROM size
+is given by `32 KiB x (2 ^ <value>)`:
```
```suggestion
This byte indicates how much ROM is present on the cartridge.
In most cases, the ROM size is given by `32 KiB × (1 << <size>)`:
```

ihonen replied:
I'm not sure the shift expression is an improvement. 🤔

ISSOtm replied:
The "power" operator like this is ambiguous with XOR. A double-star might work if you think shifting is not a good idea.

ihonen replied:
Good point, I didn't think of that. Let's go with the shift then.

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
 [^weird_rom_sizes]:
 Only listed in unofficial docs. No cartridges or ROM files using these sizes are known.
-As the other ROM sizes are all powers of 2, these are likely inaccurate. The source for these
+As the other ROM sizes are all powers of 2, these are likely inaccurate. The source of these
 values is unknown.
```
```suggestion
As the other ROM sizes are all powers of 2, these are likely inaccurate.
The source of these values is unknown.
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
-## 0149 - RAM Size
+## 0149 – RAM Size
+
+This byte indicates how large the RAM chip on the cartridge is, if any.
```
```suggestion
This byte indicates how much RAM is present on the cartridge, if any.
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
- $05 | 64 KB  | 8 banks of 8 KB each
+Code  | SRAM size | Comment
+------|----------|---------
+`$00` |   0       | No RAM [^mbc2]
+`$01` |   –       | Unused [^2kib_sram]
+`$02` |   8 KiB   |  1 bank
+`$03` |  32 KiB   |  4 banks of 8 KiB each
+`$04` | 128 KiB   | 16 banks of 8 KiB each
+`$05` |  64 KiB   |  8 banks of 8 KiB each
 
 [^mbc2]:
```
This is now covered by the paragraph above, no need to make it a footnote anymore.

```suggestion
`$00` |   0       | No RAM
`$01` |   –       | Unused [^2kib_sram]
`$02` |   8 KiB   |  1 bank
`$03` |  32 KiB   |  4 banks of 8 KiB each
`$04` | 128 KiB   | 16 banks of 8 KiB each
`$05` |  64 KiB   |  8 banks of 8 KiB each

```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
 [^mbc2]:
-When using a MBC2 chip, $00 must be specified as the RAM Size, even though
-the MBC2 includes a built-in RAM of 512 x 4 bits.
+When using a MBC2 chip, `$00` must be specified as the RAM Size, even though
+MBC2 cartridges contain an on-chip RAM of 512 x 4 bits.
```
```suggestion
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
 [^2kib_sram]:
-Listed in various unofficial docs as 2KB. However, a 2KB RAM chip was never used in a cartridge.
-The source for this value is unknown.
+Listed in various unofficial docs as 2 KiB. However, a 2 KiB RAM chip was never used in a cartridge.
```
```suggestion
Listed in various unofficial docs as 2 KiB.
However, a 2 KiB RAM chip was never used in a cartridge.
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
-Various "PD" ROMs ("Public Domain" homebrew ROMs generally tagged "(PD)"
-in the filename) are known to use the $01 RAM Size tag, but this is believed
-to have been a mistake with early homebrew tools and the PD ROMs often don't use
+Various "PD" ROMs ("Public Domain" homebrew ROMs, generally tagged with `(PD)`
+in the filename) are known to use the `$01` RAM Size tag, but this is believed
+to have been a mistake with early homebrew tools, and the PD ROMs often don't use
 cartridge RAM at all.
```
```suggestion
Various "PD" ROMs ("Public Domain" homebrew ROMs, generally tagged with `(PD)` in the filename) are known to use the `$01` RAM Size tag, but this is believed to have been a mistake with early homebrew tools, and the PD ROMs often don't use cartridge RAM at all.
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
 
-Specifies if this version of the game is supposed to be sold in Japan,
-or anywhere else. Only two values are defined.
+This byte specifies whether this version of the game is intended to be sold in
+Japan or elsewhere.
```
```suggestion
This byte specifies whether this version of the game is intended to be sold in Japan or elsewhere.
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
 
+## 014D – Header Checksum
+
+This byte contains an 8-bit checksum computed from the cartridge header bytes
+`$0134`-`$014C`. The boot ROM computes the checksum as follows:
```
```suggestion
This byte contains an 8-bit checksum computed from the cartridge header bytes `$0134`-`$014C`.
The boot ROM computes the checksum as follows:
```

ISSOtm commented on src/The_Cartridge_Header.md:
````diff
@@ -127,145 +133,150 @@ Code | Publisher
+
+```c
+uint8_t checksum = 0;
+for (uint16_t address = 0x0134; address < 0x014C; address++) {
````
```suggestion
for (uint16_t address = 0x0134; address <= 0x014C; address++) {
```
"There are two hard things in programming:"

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -127,145 +133,150 @@ Code | Publisher
 
-If the byte at $014D does not match the lower 8 bits of `x`, the boot ROM will lock up,
-and the cartridge program **won't run**.
+The boot ROM verifies this checksum. If the byte at `$014D` does not match the
+lower 8 bits of `checksum`, the boot ROM will lock up, and the program on the
+cartridge **won't run**.
```
```suggestion
The boot ROM verifies this checksum.
If the byte at `$014D` does not match the lower 8 bits of `checksum`, the boot ROM will lock up, and the program in the
cartridge **won't run**.
```

ISSOtm reviewed (changes requested):
Great! Most of my suggested changes are reformatting I'd like done while we're modifying the text, but there are also a couple of touch-ups. Once an agreement is reached on, ironically, the very last line in the document, then this PR can be merged! :)

ISSOtm commented:
To avoid dragging this PR on for longer than it already has, and to avoid blocking some upcoming changes, I'll merge this now.

ISSOtm commented:
Thanks a lot, @ihonen!

avivace reviewed (commented):
I don't know why this PR was merged when there were still major points that needed to be addressed.
The contributor difficulties with UI shouldn't put pressure on maintainers, this project is not a playground to learn git/github and although we're available to help, a draft PR may wait indefinitely if the basic interactions between contributor and maintainers cannot be carried out with enough clarity.

avivace commented:
I reverted the changes. We need to :
- Handle the points I mentioned in my last review
- Remove the `-` related changes

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,74 @@
+## 0104-0133 — Nintendo Logo
+
+This area contains a bitmap that is displayed when the Game Boy is powered on.
+They must match the following (hexadecimal) dump, otherwise [the boot ROM](<#Power-Up Sequence>) won't allow the game to run:
```
Leftover from when "This area" was "These bytes".
```suggestion
It must match the following (hexadecimal) dump, otherwise [the boot ROM](<#Power-Up Sequence>) won't allow the game to run:
```

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,74 @@
-Nintendo has reduced the length of this area to 15 characters, and some
-months later they had the fantastic idea to reduce it to 11 characters
-only. The new meaning of the ex-title bytes is described below.
+- $0104–011B encode the top half of the logo, $011C–0133 encode the bottom half of it.
```
Should addresses be written in backticks or not? This PR isn't very consistent, but the rest of Pan Docs isn't either; we should at least decide it for this PR and harmonize the rest later.

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,74 @@
+- $0104–011B encode the top half of the logo, $011C–0133 encode the bottom half of it.
+- For each half, each nibble encodes 4 pixels (the MSB corresponds to the leftmost pixel, the LSB to the rightmost); a pixel is lit if its bit is set.
+- The 4-pixel "groups" are laid out top to bottom, left to right.
+- And finally, the monochrome models upscale the entire thing by a factor of 2 (leading to somewhat chunky pixels).
```
The point was to say that this is done after all of the other transformations.

ISSOtm commented on src/The_Cartridge_Header.md:
```diff
@@ -1,65 +1,74 @@
+## 0144-0145 — New Licensee Code
+
+This area contains a two-character ASCII "licensee code" indicating the game's publisher.
+It is only meaningful if the [Old Licensee Code](<#014B — Old Licensee Code>) is exactly $33 (which is essentially all games made after the SGB was released); otherwise, the old code must be considered.
```
```suggestion
It is only meaningful if the [Old Licensee Code](<#014B — Old Licensee Code>) is exactly $33 (which is essentially the case for all games made after the SGB was released); otherwise, the old code must be considered.
```

ISSOtm commented on src/Power_Up_Sequence.md:
```diff
@@ -206,7 +206,7 @@ Rely on it at your own risk.
 
 The console's WRAM and HRAM are random on power-up.
-[Different models tend to exhibit different patterns](https://twitter.com/CasualPkPlayer/status/1409752977812852736?s=19), but they are random nonetheless, even depending on factors such as the ambient teperature.
+[Different models tend to exhibit different patterns](https://twitter.com/CasualPkPlayer/status/1409752977812852736), but they are random nonetheless, even depending on factors such as the ambient teperature.
```
```suggestion
[Different models tend to exhibit different patterns](https://twitter.com/CasualPkPlayer/status/1409752977812852736), but they are random nonetheless, even depending on factors such as the ambient temperature.
```
